pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java

```diff
@@ -106,15 +108,17 @@ public BrokerResponseNative reduceOnStreamResponse(BrokerRequest brokerRequest,
     return brokerResponseNative;
   }
 
-  private static void processIterativeServerResponse(StreamingReducer reducer, ExecutorService executorService,
+  @VisibleForTesting
+  static void processIterativeServerResponse(StreamingReducer reducer, ExecutorService executorService,
       Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> serverResponseMap, long reduceTimeOutMs,
-      ExecutionStatsAggregator aggregator) throws Exception {
+      ExecutionStatsAggregator aggregator) throws ExecutionException {
     int cnt = 0;
-    Future[] futures = new Future[serverResponseMap.size()];
-    CountDownLatch countDownLatch = new CountDownLatch(serverResponseMap.size());
-
+    Future<Void>[] futures = new Future[serverResponseMap.size()];
+    // based on ideas from on https://stackoverflow.com/questions/19348248/waiting-on-a-list-of-future
+    // mostly because we need to handle and transfer the exception in threads into caller thread.
+    ExecutorCompletionService<Void> countDownHelper = new ExecutorCompletionService<>(executorService);
```

Comment:
Another way to fix is:
in the end of this method, call
```
for (Future future : futures) {
    future.get(reduceTimeoutMs, TimeUnit.MILLISECONS);
}
```
Somehow this has the potential to multiply the time wait by `futures.length()` times, and it breaks the purpose of parallel execution.

I remember in Scala I have written or found some utility class to help with this kind of `WaitForMultipleObjects` thing, not sure where to find it now.

Reply:
wouldn't `CompletableFuture.allOf` satisfy your requirement here - https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#allOf-java.util.concurrent.CompletableFuture...- ?
Or is it too invasive change to replace `Future` with `CompletableFuture`?

Reply:
Thought about this, at first thought gRPC calls returns a `Future<Void>` and it is painful to convert to `CompletableFuture`.
But I found https://stackoverflow.com/questions/23301598/transform-java-future-into-a-completablefuture , because we are creating Future by `ExecutorService.submit()` we can switch to use `CompletableFuture.supplyAsync()` to create CompletableFuture "natively".

Tried it out and it works. Now the code is shorter and cleaner.